feat : Creating support for Event Agendas, part1 : Agenda Category CRUD

Atharva-Kanherkar opened the pull request:
**What kind of change does this PR introduce?**

 Feature 

**Issue Number:**

Fixes #1588 

**Did you add tests for your changes?**

 Yes

 

**Summary**
### 1. **Support for Agenda Categories:**
   - **Changes Introduced:** This pull request adds functionality for Agenda Categories, including CRUD operations (Create, Read, Update, Delete).
   - **Explanation:** Users can now create, view, update, and delete agenda categories. These categories are considered universal and can be used across the entire organization.

### 2. **AgendaCategory-Organization Relationship:**
   - **Introduction of a New Field:** A new field, "agendacategories," has been added to the organization schema as an array.
   - **Purpose:** This new field establishes a relationship between organizations and agenda categories. Each organization can now have an array of associated agenda categories.

### 3. **Automatic Association of New Categories:**
   - **Automation:** Any newly created categories are automatically associated with the corresponding organization.
   - **Simplifying Workflow:** This automation ensures that when users create new agenda categories, they are immediately linked to the organization they belong to, simplifying the categorization process.

 
**Does this PR introduce a breaking change?**

 No
 

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-api/blob/master/CONTRIBUTING.md)?**

 Yes

Cioppolo14 commented:
@Atharva-Kanherkar Please fix the failing tests.

palisadoes commented:
> > @Atharva-Kanherkar Please fix the failing tests.
> 
> Hello, i havent changed any of these files which codecov is giving me an error for. The files which it is telling that miss test coverage, havent even been written by me or changed. Can you please help me figure out why these fail?

Codecov sometimes includes files you fixed due to merge conflicts. Let's add some reviewers

Atharva-Kanherkar commented:
> > > @Atharva-Kanherkar Please fix the failing tests.
> > 
> > Hello, i havent changed any of these files which codecov is giving me an error for. The files which it is telling that miss test coverage, havent even been written by me or changed. Can you please help me figure out why these fail?
> 
> Codecov sometimes includes files you fixed due to merge conflicts. Let's add some reviewers

Alright! Until then, can I try making another two PRs?

palisadoes commented:
We need to get this reviewed first

EshaanAgg commented on schema.graphql:
```diff
@@ -70,6 +70,17 @@ enum AdvertisementType {
+  createdBy: User!
+  description: String
+  name: String!
+  organization: Organization
```
`Organization!`

EshaanAgg commented on src/models/AgendaCategory.ts:
```diff
@@ -0,0 +1,67 @@
+  },
+  updatedAt: {
+    type: Date,
+
```
Remove this `newline`

EshaanAgg commented on src/models/Organization.ts:
```diff
@@ -134,6 +137,12 @@ const organizationSchema = new Schema(
         required: true,
       },
     ],
+    agendaCategories: [
+      {
+        type: Schema.Types.ObjectId,
+        ref: "AgendaCategory",
+      },
+    ],
```
No need to store this here. This is already stored on `AgendaCategory`

EshaanAgg commented on src/models/AgendaCategory.ts:
```diff
@@ -0,0 +1,67 @@
+    type: String,
+    trim: true,
+  },
+  organization: {
```
Rename to `organizationId`

EshaanAgg commented on src/resolvers/AgendaCategory/organization.ts:
Create a field-level resolver for `updatedBy` in AgendaCategory. Also create field level resolver for Category on the Organization model.

EshaanAgg commented on src/resolvers/Query/getAllAgendaCategories.ts:
```diff
@@ -0,0 +1,23 @@
+import { AgendaCategoryModel } from "../../models";
+import { errors, requestContext } from "../../libraries";
+import { AGENDA_CATEGORY_NOT_FOUND_ERROR } from "../../constants";
+/* eslint-disable */
```
Why this comment?

Atharva-Kanherkar replied:
Hello, it is giving me errors for TSdocs I have researched a lot, but i do not see a reason why the lint errors are coming. The code and doc style seems correct to me. After a lot of trying, i finally added this comment.

EshaanAgg commented on src/resolvers/Query/getAllAgendaCategories.ts:
```diff
@@ -0,0 +1,23 @@
+import { errors, requestContext } from "../../libraries";
+import { AGENDA_CATEGORY_NOT_FOUND_ERROR } from "../../constants";
+/* eslint-disable */
+/**
+ * Resolver function for the GraphQL query 'agendaCategories'.
+ *
+ * This resolver fetches all agenda categories from the database.
+ *
+ * @returns {Promise<Array<AgendaCategory>>} A promise that resolves to an array of agenda categories.
+ * @throws {InternalServerError} Throws an error for potential issues during agenda categories fetching.
+ */
+/* eslint-enable */
+export const agendaCategories: QueryResolvers["agendaCategories"] =
```
What is the point of this query?

Atharva-Kanherkar replied:
I made this so that a user could see all the categories of agendas an organization has, so he can create agendaItems on that basis. It would help a person creating a new agendaItem to check if his agendaItem does match any existing category.

EshaanAgg replied:
This is but returning the same across all the organizations and not a specific one. The same should be done via a field level resolver on the Organization model.

Atharva-Kanherkar replied:
Yes you are right, i am sorry. Now I see it does not actually have a point. I will delete this and add one for organization.

EshaanAgg commented on src/resolvers/Query/getAgendaCategoryById.ts:
```diff
@@ -0,0 +1,40 @@
+import { AgendaCategoryModel } from "../../models";
+import { errors, requestContext } from "../../libraries";
+import { AGENDA_CATEGORY_NOT_FOUND_ERROR } from "../../constants";
+/* eslint-disable */
```
Why this comment?

Atharva-Kanherkar replied:
Removed all the eslint disable comments.

EshaanAgg commented on src/typeDefs/inputs.ts:
```diff
@@ -311,6 +317,12 @@ export const inputs = gql`
+  input UpdateAgendaCategoryInput {
+    name: String
+    description: String
+    organizationId: ID
```
No need to include the organizationId here as you can't possibly change it.

EshaanAgg commented on src/typeDefs/queries.ts:
```diff
@@ -19,6 +19,10 @@ export const queries = gql`
 
+    agendaCategory(id: ID!): AgendaCategory
+
+    agendaCategories: [AgendaCategory]
```
No need for this.

Atharva-Kanherkar replied:
Hello, do you mean that there is no need for these resolvers or just the typedefs?

EshaanAgg replied:
No need for the resolver.

EshaanAgg commented on src/resolvers/Query/getAgendaCategoryById.ts:
Rename file to match the query.

EshaanAgg commented on src/resolvers/Query/getAgendaCategoryById.ts:
```diff
@@ -0,0 +1,40 @@
+ */
+/* eslint-enable */
+
+export const agendaCategory: QueryResolvers["agendaCategory"] = async (
```
Should there be checks like only the members of the organization but be able to view the agenda category? @palisadoes

Atharva-Kanherkar replied:
@palisadoes  Hello, any updates on this?

EshaanAgg commented on src/typeDefs/queries.ts:
```diff
@@ -19,6 +19,10 @@ export const queries = gql`
       organizationId: ID!
     ): [ActionItemCategory]
 
+    agendaCategory(id: ID!): AgendaCategory
```
The return type should be not null.

EshaanAgg commented on src/typeDefs/types.ts:
```diff
@@ -26,6 +26,17 @@ export const types = gql`
+    _id: ID!
+    name: String!
+    description: String
+    organization: Organization
```
Organization would be not null always.

EshaanAgg commented on src/models/AgendaCategory.ts:
```diff
@@ -0,0 +1,66 @@
+  _id: Types.ObjectId; // Unique identifier for the agenda category.
+  name: string; // Name of the agenda category.
+  description?: string; // Optional description of the agenda category.
+  organization: PopulatedDoc<InterfaceOrganization & Document>; // Reference to the organization associated with the agenda category.
```
Rename to organizationId

EshaanAgg commented on src/resolvers/Mutation/createAgendaCategory.ts:
```diff
@@ -0,0 +1,91 @@
+import { adminCheck, superAdminCheck } from "../../utilities";
+import { cacheOrganizations } from "../../services/OrganizationCache/cacheOrganizations";
+import { findOrganizationsInCache } from "../../services/OrganizationCache/findOrganizationsInCache";
+/* eslint-disable */
```
Remove this eslint-diable comments

EshaanAgg commented on src/resolvers/Mutation/updateAgendaCategory.ts:
```diff
@@ -0,0 +1,100 @@
+      {
+        $set: {
+          updatedBy: context.userId,
+          // eslint-disable-next-line
```
These type assertions shouldn't be required

EshaanAgg reviewed (changes requested):
The overall changes look good, but there are certain field level resolves missing. Either add all of them in this PR, or remove them all and add them in the next PR. Also remove the linting diable comments.

Atharva-Kanherkar commented:
> The overall changes look good, but there are certain field level resolves missing. Either add all of them in this PR, or remove them all and add them in the next PR. Also remove the linting diable comments.

Hello, could you please tell me the field level resolvers which are missing? I have added `updateBy`, `createdBy`, and `organization` in the agendaCategory model. And I have added `agnendaCategories` resolver in the organization model... And ill surely try to remove the eslint disable comments. Thank you!

Atharva-Kanherkar commented:
> > The overall changes look good, but there are certain field level resolves missing. Either add all of them in this PR, or remove them all and add them in the next PR. Also remove the linting diable comments.
> 
> Hello, could you please tell me the field level resolvers which are missing? I have added `updateBy`, `createdBy`, and `organization` in the agendaCategory model. And I have added `agnendaCategories` resolver in the organization model... And ill surely try to remove the eslint disable comments. Thank you!

Removed the lint disable comments!

Atharva-Kanherkar commented:
@palisadoes Hello, the linting error which is coming here is of another file i have not authored/edited. And, can i get some assistance on fixing this issue : 
<img width="670" alt="image" src="https://github.com/PalisadoesFoundation/talawa-api/assets/142440039/e9e8e34a-6308-4a21-8c88-ee15520c1091">

SiddheshKukade commented:
@Atharva-Kanherkar you can still fix other checks. And what is the error btw ?

Atharva-Kanherkar commented:
> Please fix.

I have fixed the error.

Atharva-Kanherkar commented:
I ran prettier and tried to commit my changes, it then showed me errors in files again, i have not authored/changed...
<img width="622" alt="image" src="https://github.com/PalisadoesFoundation/talawa-api/assets/142440039/d3328c7f-697a-462c-b135-0de673f1045f">

SiddheshKukade commented:
Fix formatting errors

Atharva-Kanherkar commented:
I will fix the one failing test ASAP.

Atharva-Kanherkar commented:
@palisadoes  I ran the test cases. all of them pass : 
<img width="637" alt="image" src="https://github.com/PalisadoesFoundation/talawa-api/assets/142440039/21a6d5f9-9849-46a1-b760-e88bb08570dc">

Atharva-Kanherkar commented:
@palisadoes  Hello, are there any further changes required?

SiddheshKukade reviewed (approved):
Please check the comments. Otherwise looks good to merge.

SiddheshKukade commented on tests/resolvers/Query/actionItemsByOrganization.spec.ts:
```diff
@@ -6,16 +6,14 @@ import { actionItemsByOrganization as actionItemsByOrganizationResolver } from "
 beforeAll(async () => {
   MONGOOSE_INSTANCE = await connect();
-  [, testEvent, testOrganization] = await createTestActionItems();
+  [, , testOrganization] = await createTestActionItems();
```
Is this fine ? does it produces error ?
